File: cli-proxy-logger-py/cli_proxy_logger/recorder.py
"""Persists exchanges to a per-day JSONL file and keeps the most recent ones in
memory for the UI/API. One JSON object per line in logs/YYYY-MM-DD.jsonl.
"""

import json
import os
import threading
from datetime import datetime, timezone

MEM_LIMIT = 500


class Recorder:
    def __init__(self, config):
        self.config = config
        self.recent = []
        self._lock = threading.Lock()
        os.makedirs(config["logDir"], exist_ok=True)

    def _file(self):
        day = datetime.now(timezone.utc).strftime("%Y-%m-%d")
        return os.path.join(self.config["logDir"], f"{day}.jsonl")

    def record(self, exchange):
        with self._lock:
            self.recent.insert(0, exchange)
            if len(self.recent) > MEM_LIMIT:
                del self.recent[MEM_LIMIT:]
        try:
            with open(self._file(), "a", encoding="utf-8") as f:
                f.write(json.dumps(exchange, ensure_ascii=False) + "\n")
        except OSError as err:
            print(f"[recorder] failed to write log: {err}")

    def list(self, limit=100):
        with self._lock:
            items = list(self.recent[:limit])
        out = []
        for e in items:
            req = e.get("request") or {}
            res = e.get("response") or {}
            out.append({
                "id": e.get("id"),
                "ts": e.get("ts"),
                "durationMs": e.get("durationMs"),
                "wire": e.get("wire"),
                "method": e.get("method"),
                "url": e.get("url"),
                "model": req.get("model"),
                "stream": req.get("stream", False),
                "resStatus": e.get("resStatus"),
                "toolCallNames": [t.get("name") for t in (res.get("toolCalls") or [])],
                "error": e.get("error"),
            })
        return out

    def get(self, ex_id):
        with self._lock:
            for e in self.recent:
                if e.get("id") == ex_id:
                    return e
        return None

    def clear(self):
        # Clears the in-memory list (what the UI shows). The on-disk JSONL logs
        # are left untouched — they are the durable audit trail. Returns how many
        # in-memory entries were removed.
        with self._lock:
            n = len(self.recent)
            self.recent = []
        return n

File: cli-proxy-logger/src/recorder.js
// Persists exchanges to a per-day JSONL file and keeps the most recent ones in
// memory for the UI/API. One JSON object per line in logs/YYYY-MM-DD.jsonl.

import fs from 'node:fs';
import path from 'node:path';

const MEM_LIMIT = 500;

export class Recorder {
  constructor(config) {
    this.config = config;
    this.recent = [];
    fs.mkdirSync(config.logDir, { recursive: true });
  }

  _file() {
    const day = new Date().toISOString().slice(0, 10);
    return path.join(this.config.logDir, `${day}.jsonl`);
  }

  record(exchange) {
    this.recent.unshift(exchange);
    if (this.recent.length > MEM_LIMIT) this.recent.length = MEM_LIMIT;
    try {
      fs.appendFileSync(this._file(), JSON.stringify(exchange) + '\n');
    } catch (err) {
      console.error('[recorder] failed to write log:', err.message);
    }
  }

  list(limit = 100) {
    return this.recent.slice(0, limit).map((e) => ({
      id: e.id,
      ts: e.ts,
      durationMs: e.durationMs,
      wire: e.wire,
      method: e.method,
      url: e.url,
      model: e.request?.model ?? null,
      stream: e.request?.stream ?? false,
      resStatus: e.resStatus,
      toolCallNames: (e.response?.toolCalls || []).map((t) => t.name),
      error: e.error ?? null,
    }));
  }

  get(id) {
    return this.recent.find((e) => e.id === id) || null;
  }

  // Clears the in-memory list (what the UI shows). The on-disk JSONL logs are
  // left untouched — they are the durable audit trail. Returns how many
  // in-memory entries were removed.
  clear() {
    const n = this.recent.length;
    this.recent = [];
    return n;
  }
}
